File: Java Implementation/Data Structures(java)/src/main/java/com/MathLingCS/Stack/Stack.java
package com.MathLingCS.Stack;

@SuppressWarnings("unused")

// An abstract stack interface.
public interface Stack<T> {
    /**
     * Adds an element to the top of the stack.
     * @param element the element to be added
     */
    void push(T element);

    /**
     * Removes the element at the top of the stack and returns it.
     * @return the element at the top of the stack
     */
    T pop();

    /**
     * Returns the element at the top of the stack without removing it.
     * @return the element at the top of the stack
     */
    T peek();

    /**
     * Returns the number of elements in the stack.
     * @return the number of elements in the stack
     */
    int size();

    /**
     * Returns whether the stack is empty or not.
     * @return true if the stack is empty, false otherwise
     */
    boolean isEmpty();

    /**
     * Removes all elements from the stack.
     */
    void clear();
    
}

File: Linear_Structures/Stack/Stack.py
from abc import ABC, abstractmethod
from typing import TypeVar, Generic, Iterator

T = TypeVar('T')


class Stack(ABC, Generic[T]):
    """
    Abstract class for stacks.
    """

    @abstractmethod
    def push(self, item: T) -> None:
        """
        Pushes an item to the stack.
        """
        pass

    @abstractmethod
    def pop(self) -> T:
        """
        Pops an item from the stack.
        """
        pass

    @abstractmethod
    def peek(self) -> T:
        """
        Returns the top item from the stack.
        """
        pass

    @abstractmethod
    def is_empty(self) -> bool:
        """
        Returns True if the stack is empty, False otherwise.
        """
        pass

    @abstractmethod
    def __len__(self) -> int:
        """
        Returns the number of items in the stack.
        """
        pass

    @abstractmethod
    def __contains__(self, item0: T) -> bool:
        """
        Checks if an item is in the stack.
        """
        pass

    @abstractmethod
    def clear(self) -> None:
        """
        Removes all items from the stack.
        """
        pass
